ERCS/erc-6538.md

```diff
@@ -27,65 +27,61 @@ This contract defines an `ERC5564Registry` that stores the stealth meta-address
   /// @dev MUST return zero if a registrant has not registered keys for the given inputs.
-  mapping(bytes => mapping(uint256 => bytes)) public stealthMetaAddressOf;
+  mapping(address => mapping(uint256 => bytes)) public stealthMetaAddressOf;
+  mapping(address user => uint256 nonce) public nonceOf;
+  bytes32 private constant TYPE_HASH =
+    keccak256("EIP712Domain(string name,string version,uint256 chainId,address registryContract)");
+
 
   /// @notice Sets the caller's stealth meta-address for the given stealth address scheme.
   /// @param scheme An integer identifier for the stealth address scheme.
   /// @param stealthMetaAddress The stealth meta-address to register.
   function registerKeys(uint256 scheme, bytes memory stealthMetaAddress) external {
-    stealthMetaAddressOf[abi.encode(msg.sender)][scheme] = stealthMetaAddress;
+    stealthMetaAddressOf[msg.sender][scheme] = stealthMetaAddress;
   }
 
   /// @notice Sets the `registrant`s stealth meta-address for the given scheme.
   /// @param registrant Recipient identifier, such as an ENS name.
-  /// @param scheme An integer identifier for the stealth address scheme.
+  /// @param schemeId An integer identifier for the stealth address scheme.
   /// @param signature A signature from the `registrant` authorizing the registration.
   /// @param stealthMetaAddress The stealth meta-address to register.
   /// @dev MUST support both EOA signatures and EIP-1271 signatures.
   /// @dev MUST revert if the signature is invalid.
   function registerKeysOnBehalf(
     address registrant,
-    uint256 scheme,
+    uint256 schemeId,
     bytes memory signature,
     bytes memory stealthMetaAddress
   ) external {
-    // TODO If registrant has no code, spit signature into r, s, and v and call `ecrecover`.
-    // TODO If registrant has code, call `isValidSignature` on the registrant.
-  }
-
-  /// @notice Sets the `registrant`s stealth meta-address for the given scheme.
-  /// @param registrant Recipient identifier, such as an ENS name.
-  /// @param scheme An integer identifier for the stealth address scheme.
-  /// @param signature A signature from the `registrant` authorizing the registration.
-  /// @param stealthMetaAddress The stealth meta-address to register.
-  /// @dev MUST support both EOA signatures and EIP-1271 signatures.
-  /// @dev MUST revert if the signature is invalid.
-  function registerKeysOnBehalf(
-    bytes memory registrant,
-    uint256 scheme,
-    bytes memory signature,
-    bytes memory stealthMetaAddress
-  ) external {
-    // TODO How to best generically support any registrant identifier / name
-    // system without e.g. hardcoding support just for ENS?
+    bytes32 digest = _hashTypedDataV4(
+      keccak256(
+        abi.encode(TYPE_HASH, registrant, schemeId, stealthMetaAddress, nonceOf[registrant]++)
+      )
+    );
+    require(isValidSignatureNow(registrant, digest, signature), "Invalid signature");
```

Comment:
This method is not defined anywhere in this contract. I think this is an OZ method, but I'd suggest avoiding pulling in their dependency and keeping this as a single flat solidity file. We can copy code out of that dep as needed, but currently this ERC now does not fully specify the contract